File: java/src/main/java/datastructures/DynamicArray.java
package datastructures;

/**
 * package java.util.ArrayList equivalent
 * @param <T>
 */
public class DynamicArray<T> {

    public static final Object EMPTY = null;
    private Object[] data;
    private int size;
    private int initialCapacity;

    public DynamicArray(final int initialCapacity) {
        this.data = new Object[initialCapacity];
        this.initialCapacity = initialCapacity;
        this.size = 0;
    }

    /**
     * O(1)
     *
     *
     * @param index
     * @return
     */
    public T get(final int index) {
        return (T) data[index];
    }

    /**
     * O(1)
     *
     *
     * @param index
     * @param value
     */
    public void set(final int index, final T value) {
        data[index] = value;
        size++;
    }

    /**
     * O(n)
     *
     *
     * @param index
     * @param value
     */
    public void insert(final int index, final T value) {
        // Check size
        if (size == initialCapacity) {
            resize(); // --> O(n)
        }

        // Manual Array Copy | Shift
        for (int j = size; j > index; j--) {  //--> O(n)
            data[j] = data[j - 1];
        }
        //if (size - index >= 0) System.arraycopy(data, index, data, index + 1, size - index);

        // Insert
        data[index] = value;  //--> O(1)
        size++;
    }

    public int getSize() {
        return this.size;
    }

    // Some array implementations offer methods like insert (also known as pushBack).
    // Which is like insert, only in does the insert at the end of the array.(append)
    public void add(final T value) {
        if (size == initialCapacity) {
            resize();
        }

        data[size] = value;
        size++;
    }

    /**
     * O(n)
     *
     *
     * @param index
     */
    public void delete(final int index) {
        // Copy down
        for (int j = index; j < size - 1; j++) { // --> O(n)
            data[j] = data[j + 1];
        }

        // Clear if last element in array
        if (index == size) {
            data[index] = EMPTY;
        }else {
            //Clear last value index
            data[size - 1] = EMPTY;
        }
        size--;
    }

    /**
     * O(1)
     *
     * @return
     */
    public boolean isEmpty() {
        return size == 0;
    }

    /**
     * 0(n)
     *
     * @param value
     * @return
     */
    public boolean contains(final T value) {
        for (int i = 0; i <= size - 1; i++) { // --> 0(n)
            final T currentValue = (T) data[i];
            if (currentValue.equals(value)) {
                return true;
            }
        }
        return false;
    }

    /**
     * O(n) : For loop array copy
     *
     * System.arraycopy
     *
     * http://hg.openjdk.java.net/jdk/jdk11/file/1ddf9a99e4ad/src/hotspot/share/prims/jvm.cpp
     *
     * http://hg.openjdk.java.net/jdk7/jdk7/hotspot/file/9b0ca45cd756/src/share/vm/oops/objArrayKlass.cpp#l168
     */
    private void resize() {
        final var newCapacity = initialCapacity * 2;
        final var newData = new Object[newCapacity];

        if (initialCapacity >= 0) {
            System.arraycopy(data, 0, newData, 0, initialCapacity);
        }

        //manual array copy
        /*for (int i = 0; i < initialCapacity; i++) {
            newData[i] = data[i];
        }*/

        data = newData;
        initialCapacity = newCapacity;
    }
}

File: go/datastructures/dynamic_array.go
package datastructures

type DynamicArray struct {
	size int
	initialCapacity int
	data []interface{}
}

func (da *DynamicArray) New(initialCapacity int) DynamicArray {
	return DynamicArray{size: 0, initialCapacity: initialCapacity, data: make([]interface{}, initialCapacity)}
}

func (da *DynamicArray) Get(index int) interface{} {
	return da.data[index]
}

func (da *DynamicArray) Set(index int, data interface{}) {
	da.data[index] = data
	da.size++
}

func (da *DynamicArray) Insert(index int, value interface{})  {
	// Check size
	if da.size == da.initialCapacity {
		resize(da)
	}

	// Copy up | Shift
	for j := da.size; j > index; j-- {
	    da.data[j] = da.data[j - 1]
	}

	// Insert
	da.data[index] = value
	da.size++
}

//Add is like a Append
func (da *DynamicArray) Add(value interface{})  {
	// Check size
	if da.size == da.initialCapacity {
		resize(da)
	}

	// Insert
	da.data[da.size] = value
	da.size++
}

//Slice appending
func (da *DynamicArray) Append(value interface{})  {
	da.data = append(da.data, value)
}

func (da *DynamicArray) Delete(index int) {
	// Copy down
	for  j := index; j < da.size - 1; j++ {
		da.data[j] = da.data[j + 1]
	}

	// Clear if last element in array
	if index == da.size {
		da.data[index] = nil
	}else {
		//Clear last value index
		da.data[da.size - 1] = nil
	}

	da.size--
}

func (da *DynamicArray) Contains(value interface{}) bool{
	for i := 0; i < da.size; i++ {
		if da.data[i] == value {
			return true
		}
	}
	return false
}

func (da DynamicArray) Empty() bool {
	return da.size == 0
}

func resize(da *DynamicArray) {
	newCapacity := da.initialCapacity * 2
	newData := make([]interface{}, newCapacity)

	if da.initialCapacity > 0 {
		copy(newData, da.data)
	}
	da.data = newData
	da.initialCapacity = newCapacity
}
